### PR TITLE
[5.1] Upgraded Laravel's usage of the AWS SDK for PHP

### PR DESCRIPTION
**Pull request by jeremeamia**
The Queue and Mail components have had their usage of the AWS SDK for PHP upgraded such that it works with the new [Version 3](http://blogs.aws.amazon.com/php/post/Tx29HZZVCASVU21/Version-3-of-the-AWS-SDK-for-PHP) of the SDK and the older Version 2. The version constraint used for the SDK is now `~2.4|~3.0`.

This PR along with #8918 make it so the following dependency closures are all allowed:
- Laravel 5.1 + Guzzle 6 + AWS SDK 3
- Laravel 5.1 + Guzzle 5 + AWS SDK 3
- Laravel 5.1 + Guzzle 6 + Guzzle 3 + AWS SDK 2
- Laravel 5.1 + Guzzle 5 + Guzzle 3 + AWS SDK 2

I added a unit test in the Mail component and modified some in the Queue component. Unit tests all pass when using Version 2.x and Version 3.x of the AWS SDK. I did not do any manual tests via the framework, since I am not currently setup to do that. You might consider doing this if you have some code that exercises the SES or SQS features, and you need additional peace of mind.

### CODE REVIEW

**Comment by jeremeamia** [comment]
Tests are failing on PHP 5.4. Version 3 of the SDK requires PHP 5.5, and since "minimum-stability" is set to "dev", it's trying to use  `aws/aws-sdk-php:3.0.0-beta.1`, which only required PHP 5.4. The beta version definitely will not work with this codebase, but I'm not sure how to get around this. Any ideas @taylorotwell or @GrahamCampbell?

**Comment by taylorotwell** [comment]
Do we need to revisit he minimum version requirement for Laravel 5.1
perhaps?

**Comment by crynobone** [comment]
Shouldn't it fallback to a compatible version (aws/aws-sdk-php:2.*) when testing on PHP 5.4?

**Comment by jeremeamia** [comment]
> Shouldn't it fallback to a compatible version (aws/aws-sdk-php:2.*) when testing on PHP 5.4?

@crynobone It should, but it's not because `minimum-stability` is set to `dev`, which is causing the beta to be used. If it was set to stable (or unset), then it would fallback to 2.x as expected.

> Do we need to revisit he minimum version requirement for Laravel 5.1 perhaps?

@taylorotwell Not really, it's just `minimum-stability` causing problems here. That is going away at some point, I assume, right?

**Comment by franzliedke** [comment]
Is there a stable equivalent for the `@dev` flag that you can apply to package version constraints?

**Comment by barryvdh** [comment]
Yeah, I think it's just `@stable` ;)

File: composer.json [hunk 1]
```diff
@@ -101,7 +101,7 @@
         }
     },
     "suggest": {
-        "aws/aws-sdk-php": "Required to use the SQS queue driver and SES mail driver (~2.4).",
+        "aws/aws-sdk-php": "Required to use the SQS queue driver and SES mail driver (~2.4|~3.0).",
```
Comment (GrahamCampbell) [inline composer.json]: This needs updating in our other composer.json file too. ;)

**Comment by GrahamCampbell** [comment]
I think we should change laravel's tests to use stable packages only anyway. I suggested this a year ago.

**Comment by barryvdh** [comment]
But Symfony 2.7 isn't released yet, so must be dev.

**Comment by GrahamCampbell** [comment]
> But Symfony 2.7 isn't released yet, so must be dev.

But when it is released, we can change that. :)

**Comment by GrahamCampbell** [comment]
And for L5.2, I think the plan is to drop guzzle 5 and aws sdk v2, so it can be save for us to set that to dev with prefer stable during the development cycle.

**Comment by barryvdh** [comment]
Yeah, but only if you're going to drop 5.4 support.

**Comment by GrahamCampbell** [comment]
> Yeah, but only if you're going to drop 5.4 support.

Yeh, we're moving to 5.5.9+ in L5.2.

**Comment by GrahamCampbell** [comment]
Change of plan. Laravel 5.1 is 5.5.9+ now. :)

**Comment by GrahamCampbell** [comment]
Thanks so much @jeremeamia! I'll finish this off from here. :)

**Comment by jeremeamia** [comment]
:+1:

**Comment by franzliedke** [comment]
> Change of plan. Laravel 5.1 is 5.5.9+ now. :)

That's a pretty far-reaching change so shortly before the release... :-(

**Pull request closed without merging**